xmodule/library_content_block.py

```diff
@@ -251,6 +234,64 @@ def _publish_event(self, event_name, result, **kwargs):
         self.runtime.publish(self, f"edx.librarycontentblock.content.{event_name}", event_data)
         self._last_event_result_count = len(result)  # pylint: disable=attribute-defined-outside-init
 
+    @classmethod
+    def _get_valid_children(cls, library_children, candidates, manual):
+        """
+        Dynamically selects block_ids which children are possible for selection
+        """
+        if candidates and manual:
+            return candidates
+        return {(child.block_type, child.block_id) for child in library_children}
+
+    @classmethod
+    def make_selection(cls, selected, library_children, candidates, max_count, manual, shuffle):
+        """
+        Dynamically selects block_ids indicating which of the possible children are displayed to the current user.
+        The blocks returned are kept consistent for a user,
+        unless changes have been made to the library's contents or the settings of the block.
+        Returns:
+            A dict containing the following keys:
+            'selected' (set) of (block_type, block_id) tuples assigned to this student
+            'invalid' (set) of dropped (block_type, block_id) tuples that are no longer valid
+            'overlimit' (set) of dropped (block_type, block_id) tuples that were previously selected
+            'added' (set) of newly added (block_type, block_id) tuples
+        """
+        old_selected = set(tuple(k) for k in selected)
+        valid_block_keys = LibraryContentBlock._get_valid_children(library_children, candidates, manual)
+        valid_old_block_keys = valid_block_keys.intersection(old_selected)
+        valid_additions = valid_block_keys.difference(valid_old_block_keys)
+        new_selected = set()
+        overlimit_block_keys = []
+        size = max_count if len(valid_block_keys) >= max_count else len(valid_block_keys)
+
+        if shuffle:
+            #determine how many blocks need to be added or removed.
+            vaccancies_in_selected = (
+                size - len(valid_old_block_keys)
+                if size >= 0
+                else len(valid_block_keys) - len(valid_old_block_keys)
+            )
+
+            if vaccancies_in_selected < 0:
+                new_selected = list(valid_old_block_keys)[0:size]
+                overlimit_block_keys = list(valid_old_block_keys)[size:]
+            else:
+                if not manual:
+                    new_selected = list(valid_old_block_keys) + random.sample(valid_additions, vaccancies_in_selected)
+                else:
+                    new_selected = list(valid_old_block_keys) + list(valid_additions)[0:vaccancies_in_selected]
+                if new_selected != old_selected:
+                    random.shuffle(new_selected)
+        else:
+            new_selected = list(valid_block_keys)[:size]
```

Comment:
this would pick the first N blocks out of the library every single time, which ain't right.

instead, N blocks should be randomly picked from the library for each learner, and if `shuffle` is enabled, then the order of them should be randomized.

Reply:
I'm confused by `this` here. 

if this refers to line 286:
note that line 286 provides selection for the case when shuffle = false.

if `this` refers to lines 267-284:

Note we need to preserve `selected` as much as possible, not across users, but rather for the same user. in cases where blocks are removed from the library, or max_count is decreased, we need to keep learners with all possible valid questions they have seen before. This way, updates to the block on publish will not vary significantly. The randomization from sample and shuffle is still included.